### ISSUE TITLE
Curried functions being mangled via DU, List.fold and match combination

### ISSUE THREAD

**Issue by davedawkins** [fable-compiler/Fable#2356]
### Description

Trying to find the "In x" value in "[ In f1; Out f2]" returns garbage, when f1 is a curried function and List.fold is used for the search. 

### Repro code

```fs

    type Fn = bool -> int -> string
    type Thing =
        | In of Fn
        | Out of Fn

    let findThing (things:Thing list) =
        let folder (a : Fn option) t =
            match t with
            | In x -> Some x // Found an In, set accumulator
            | _ -> a // pass accumulator through unchanged

        things |> List.fold folder None  // Searching for an "In x"

    let testData = [ In (fun b i -> "fly");  Out (fun b i -> "fade")]

    let test = match findThing testData with
                        | Some f -> f true 1
                        | None -> "nothing"

    console.log($"{test}") // Expect this to write "fly" to the console

```

Any attempt to simplify this case makes the bug go away. I've simplified as much as I can.

Interestingly, if you search for the "Out x", which is last in the list, then this succeeds. It's as if the found value doesn't like being passed back through another iteration of the fold as the accumulator value.

This code has been verified as working in Interactive Notebook

### Expected and actual results

I expect this to print "fly", instead it prints "(a2) => f(a1, a2)"

### Related information

```
david@davebook Sutil % dotnet fable --version
Fable: F# to JS compiler 3.0.5
Thanks to the contributor! @alfonsogarciacaro
3.0.5

david@davebook Sutil % uname -a
Darwin davebook.local 20.2.0 Darwin Kernel Version 20.2.0: Wed Dec  2 20:39:59 PST 2020; root:xnu-7195.60.75~1/RELEASE_X86_64 x86_64
```

**Comment by davedawkins** [comment]
This version of the find function works for me, as a rough test (using for-loop instead of fold)
```fs

let private findTransition2 (intro:bool) (trans : TransitionAttribute list) : TransitionBuilder option =
    let mutable result : TransitionBuilder option = None
    for x in trans do
        if result.IsNone then
            result <- match x with
                        | In x -> if intro then Some x else None
                        | Out x -> if not intro then Some x else None
                        | _ -> None
    result
```

**Comment by alfonsogarciacaro** [comment]
Thanks a lot for isolating the code to reproduce the problem! Yes, there are some similar issues reported with the "uncurrying" label reported several months ago, I need to fix this 😅 Right now the uncurrying is applied when passing a function as argument or put in a record field, I think we also need to check other cases like wrapping the function in a DU (as in here) or in a list/array/tuple/option.

**Comment by alfonsogarciacaro** [comment]
Should be fixed in 3.1.3 :+1:
